extensions/amp-story-shopping/0.1/amp-story-shopping-config.js

```diff
@@ -1,19 +1,152 @@
+    'aggregateRating': [validateRequired, validateObject],
+    /* Optional Attrs */
+    'productColor': [validateString],
+    'productSize': [validateString],
+    'productIcon': [validateURLs],
+    'productTagText': [validateString],
+    'reviewsData': [validateURLs],
+    'ctaText': [validateNumber],
+    'shippingText': [validateNumber],
+  },
+};
+
+const essentialFields = [
+  'productPrice',
+  'productTitle',
+  'productImages',
+  'productBrand',
+  'productId',
+];
+
 /**
  * @typedef {{
  *  items: !Array<!ShoppingConfigDataDef>,
  * }}
  */
 let ShoppingConfigResponseDef;
 
+/**
+ * Validates an Object using its field name as a key for one of the validation objects above.
+ * @param {string} field
+ * @param {?Object=} value
+ */
+function validateObject(field, value = undefined) {
+  validateConfig(value, VALIDATION_OBJECTS[field]);
+}
+
+/**
+ * Validates if a required field exists for shopping config attributes
+ * @param {string} field
+ * @param {?string=} value
+ */
+function validateRequired(field, value = undefined) {
+  if (value === undefined) {
+    throw Error(`Field ${field} is required.`);
+  }
+}
+
+/**
+ * Validates if string type for shopping config attributes
+ * @param {string} field
+ * @param {?string=} str
+ */
+function validateString(field, str = undefined) {
+  if (typeof str !== 'string') {
+    throw Error(`${field} ${str} is not a string.`);
+  }
+}
+
+/**
+ * Validates number in shopping config attributes
+ * @param {string} field
+ * @param {?number=} number
+ */
+function validateNumber(field, number = undefined) {
+  if (typeof number !== 'number') {
+    throw Error(`Value ${number} for field ${field} is not a number`);
+  }
+}
+
+/**
+ * Validates url of shopping config attributes
+ * @param {string} field
+ * @param {?Array<string>=} url
+ */
+function validateURLs(field, url = undefined) {
+  if (url === undefined) {
+    return;
+  }
+
+  const urls = Array.isArray(url) ? url : [url];
+
+  urls.forEach((url) => {
+    assertHttpsUrl(url.url ?? url, `amp-story-shopping-config ${field}`);
+  });
+}
+
+/**
+ * Validates the shopping config of a single product.
+ * @param {!ShoppingConfigDataDef} shoppingConfig
+ * @param {!Object<string, !Array<function>>} validationObject
+ * @return {boolean}
+ */
+export function validateConfig(
+  shoppingConfig,
+  validationObject = VALIDATION_OBJECTS['productValidationConfig']
+) {
+  let isValidConfig = true;
+
+  Object.keys(validationObject).forEach((configKey) => {
+    const validationFunctions = validationObject[configKey];
+    validationFunctions.forEach((fn) => {
+      try {
+        /* This check skips optional attribute validation */
+        if (
+          shoppingConfig[configKey] !== undefined ||
+          validationFunctions.includes(validateRequired)
+        ) {
+          fn(configKey, shoppingConfig[configKey]);
+        }
+      } catch (err) {
+        if (essentialFields.includes(configKey)) {
```

Comment:
How is `essentialFields` different than `validateRequired`?

Reply:
If a shopping tag is missing something in `essentialFields` will block rendering of the shopping tag and not display it in the PLP. It will also throw a console error.

If a shopping tag is missing something in `validateRequired ` the shopping tag will still render, but only a console error will be shown.

Reply:
I'll add it in as a comment in the code in the next commit - Thanks @processprocess

Reply:
It feels a bit weird to be calling these required fields when they are not actually required. If we are only warning people will ship these configs and ignore the warnings. Are we sure thats what we want? If so, maybe we should rename to something like `warnIfMissing` instead.

Reply:
I find it a bit strange too. A field is either optional, or not. I don't see the value of a warning for something "required" when it isn't.

I would additionally suggest differentiating errors from warnings at the validator function result. Return it rather throwing an error. The idea with using functions is that they're composable in a single object, rather than spilling into something like `essentialFields`.

For example:

```js
return [ValidationErrorLevel.ERROR, 'foo is required'];

return [ValidationErrorLevel.WARNING, 'foo should be string'];
```

Then I'd be curious about what's done with either level. It seems like what we'd like to do for an `ERROR` is to throw at the end to block rendering, and that `WARN` would remove the validated field from the result.

Reply:
OK, I'll go with this level approach, i.e. throwing error level functions in essentialFields for rendering, but warnings for non essential ones.
As for what will happen, please see here: we will remove rendering of the product from the shopping tag and the PLP if one of these essential fields is missing. (if all essential fields are there (but there is still a field missing), a warning will be thrown, but the product will still be there.)

https://github.com/ampproject/amphtml/pull/37474#discussion_r803100775

Reply:
@processprocess I'm a bit lost here too. If the essential fields are the only ones that are "required" for redering the shopping tag and PLP, then wouldn't it make sense to make `productPrice`, `productTitle`, `productImages` (at least one), `productBrand`, `productId` the required fields and make everything else optional? [https://github.com/ampproject/amphtml/issues/36460](url)

Reply:
We will not be using essential fields in this current PR as further discussion is needed.
Please see https://github.com/ampproject/amphtml/pull/37474#discussion_r815542683 for the full explanation/to continue the discussion. Thanks.